protocol/storage/prover.go

```diff
@@ -74,36 +88,79 @@ func (sm *Prover) CalculatePoSt(ctx context.Context, start, end *types.BlockHeig
 	for i, input := range inputs {
 		commRs[i] = input.CommR
 	}
-	proofs, faults, err := sm.calculator.CalculatePost(proofs.NewSortedCommRs(commRs...), seed)
+	proofs, faults, err := p.calculator.CalculatePost(proofs.NewSortedCommRs(commRs...), seed)
 	if err != nil {
-		return nil, errors.Wrap(err, "failed to generate PoSt")
+		return nil, errors.Wrap(err, "failed to calculate PoSt")
 	}
 
-	// Compute fees.
 	if len(faults) != 0 {
 		log.Warningf("some faults when generating PoSt: %v", faults)
 		// TODO: include faults in submission https://github.com/filecoin-project/go-filecoin/issues/2889
 	}
 
-	height, err := sm.chain.ChainHeight()
+	// Compute fees.
+	balance, err := p.chain.WalletBalance(ctx, p.workerAddress)
 	if err != nil {
-		// TODO: what should happen in this case?
-		return nil, errors.Errorf("failed to submit PoSt, as the current block height can not be determined: %s", err)
+		return nil, errors.Wrapf(err, "failed to check wallet balance for %s", p.workerAddress)
 	}
-	if height.LessThan(start) {
-		// TODO: what to do here? not sure this can happen, maybe through reordering?
-		return nil, errors.Errorf("PoSt generation time took negative block time: %s < %s", height, start)
 
+	height, err := p.chain.ChainBlockHeight()
+	if err != nil {
+		return nil, errors.Wrap(err, "failed to check chain height")
+	}
+	if height.LessThan(start) {
+		return nil, errors.Errorf("chain height %s is before proving period start %s, abandoning proof", height, start)
 	}
 
-	if height.GreaterEqual(end) {
-		// TODO: pay late fees https://github.com/filecoin-project/go-filecoin/issues/2942
-		return nil, errors.Errorf("PoSt generation was too slow height=%s end=%s", height, end)
+	feeDue, err := p.calculateFee(ctx, height, end)
+	if err != nil {
+		return nil, err
+	}
+	if feeDue.GreaterThan(balance) {
+		log.Warningf("PoSt fee of %s exceeds available balance of %s for owner %s", feeDue, balance, p.workerAddress)
+		// Submit anyway, in case the balance is topped up before the PoSt message is mined.
 	}
 
 	return &PoStSubmission{
 		Proofs:   proofs,
-		Fee:      types.ZeroAttoFIL,
+		Fee:      feeDue,
 		GasLimit: types.NewGasUnits(submitPostGasLimit),
 	}, nil
 }
+
+// challengeSeed returns the PoSt challenge seed for a proving period.
+func (p *Prover) challengeSeed(ctx context.Context, periodStart *types.BlockHeight) (types.PoStChallengeSeed, error) {
+	bytes, err := p.chain.ChainSampleRandomness(ctx, periodStart)
+	if err != nil {
+		return types.PoStChallengeSeed{}, err
+	}
+
+	seed := types.PoStChallengeSeed{}
+	copy(seed[:], bytes)
+	return seed, nil
+}
+
+// calculateFee calculates any fees due with a proof submission due to faults or lateness.
+func (p *Prover) calculateFee(ctx context.Context, height *types.BlockHeight, end *types.BlockHeight) (types.AttoFIL, error) {
```

Comment:
BLOCKING: This method is collecting various elements of miner actor state (ProvingPeriodEnd, ActiveCollateral, SectorSize) and then calling static methods on the miner actor (GenerationAttackTime, LatePoStFee) to compute the late fee. This really should be a single query method on the miner.Actor. The miner actor already has this logic implemented in a similar fashion.

Consolidation of consensus logic in the actors will prevent potential consensus bugs and also simplify the model for change as these algorithms change.

Reply:
I see what you're saying, but I think there's a reasonable separation to be made somewhere here. Firstly, the *actual* fee calculation code is in the actor (factored out as a static function) and shared. This method is invoking the same calculation but not with the same parameters: the height/lateness is a hypothetical, and a comment indicates that we might also want the collateral to be a different value too.

For reasons you cited in #2973, I don't want to add an exported actor method to do this. The decision about how much buffer to add to lateness and/or collateral is not part of consensus.

Reply:
We discussed this in person. I'm intending to move the `LatePoStFee` function from a free function in the miner actor to be a method instead. This turns out to be a bit of a pain though; if I follow existing patterns it requires adding it to porcelain and probably changing the parameters, since it will necessarily be a function of a specific miner instance. I might try to find a less verbose pattern than adding every actor query method to porcelain.

Reply:
It would be my preference __not__ to add all query methods to porcelain. I believe we have several internal query method calls that don't have porcelain wrappers. The requirement for adding a porcelain method really should be that it's user facing and there is a command that calls it. That goal has gotten muddy in the porcelain implementation.

It appears that the new `getPledgeCollateralRequirement` is currently only used to provide a parameter to feed a parameter back to the late fee calculation. I would replace this function with an actor method named something like 'computeLateFeeAtBlockHeight' and have it pull the collateral requirement from state. This leaks fewer details about the fee calculation.